### ISSUE TITLE
Reaction Wheels - Determine how to control the reaction wheels from STM32 using C/C++

### ISSUE THREAD

**Issue by Aidan-Sheedy** [queens-satellite-team/adcs#49]
# Motivation/Description ⛰️
The current Reaction Wheel design needs to be reviewed see its current state. This is to ensure nothing is being missed and the project is on track. Research should also be done for the same reason - making sure nothing in the design is missing.

> It is expected that some parts of the design might not be fully understood until working with the reaction wheels directly.

# Next Steps 🚶 
- [ ] Review the reaction wheel design. Specifically, consider:
    - The electrical characteristics of the system, ie what circuitry is needed to run the reaction wheels, what connections need to be made from wheels->controller board->ADCS MCU and vice versa, etc.
    - The physical characteristics of the system, ie how the system is mounted to the chassis
- [ ] Research the reaction wheels, driver circuits, etc. and compare results with the current design.
- [ ] Create a ticket for missing design considerations, as well as extraneous parts of the design that are no longer needed.
- [ ] Update all current documentation in the formal documentation in Teams. This should incorporate all parts of the system design that do not need changes. Anything missing should have a ticket already to be addressed.

# Supporting Material and Resources 📖 
- [Maxon Reaction Wheel Guide](https://queensuca.sharepoint.com/teams/GROUP-QSET/_layouts/15/Doc.aspx?sourcedoc={9fb96dfc-b12f-4372-a0a4-f91bb2467e9c}&action=edit&wd=target%28Satellite%20ADCS.one%7C1ca96692-dfb3-460b-88e6-898f0a9a2535%2FMaxon%20Reaction%20Wheel%20Guide%7C5212604b-afdd-421f-ac9c-da0e617bb9c5%2F%29&wdorigin=NavigationUrl)
- [Manufacturer's website](https://www.maxongroup.com/maxon/view/product/397172)
- [Reaction Wheel Documentation](https://www.maxongroup.com/maxon/view/product/397172)
- [Old Teams Files](https://queensuca.sharepoint.com/:f:/r/teams/GROUP-QSET/Shared%20Documents/Satellite-ADCS/QSET%20ADCS%202021-2022/Reaction%20Wheels?csf=1&web=1&e=DAxuIp)

# Definition of Done 💯 
This task is complete when tickets have been created for all aspects of the reaction wheel design that are needed to be changed, and formal documentation has been updated with the current design.

### Tags
> tag people that should know about this issue

cc: @SeanTedesco @emmapaczkowski @Aidan-Sheedy

**Comment by anthonysteel** [comment]
Hey, I am an old team member. I actually did the purchasing for these Maxon's. Just a heads up, I read the QSET Design Book document and it seems that someone was getting a "Vcc Undervoltage error, power supply too low" @ 12VDC, which makes sense. The motor is 24V nominal operation, so try running VCC @ 24VDC. 12VDC is pretty close to the 8VDC minimum operating voltage for that board.

Also these were purchased before [ODrive's](https://odriverobotics.com/shopfolder) were popular. I have used these a bunch since graduating. They are easy to setup and can be programmed in Python. If the Maxon controller are giving you too much of a problem, I would consider buying a ODrive 3.6 with a CUI incremental encoder.

**Comment by Aidan-Sheedy** [comment]
Hey thanks @anthonysteel, that's some really good info. How many reaction wheels can each of those ODrive boards drive? It kind of looks like there's 2 each.

@SeanTedesco and @JacobORiordanRoss, do we have any options for getting a higher than 12V rail from EPS? If not, we'll need to see if we can do it on the ADCS side.

**Comment by anthonysteel** [comment]
Yeah, the Odrive 3.6 supports 2 motors on each board, so you would need 2 boards if you are planning to drive 4 motors. 
However, I would buy at least 4 if there is money in the budget. Playing with motors, you are likely to fry the controller electronics at least once (I know I have). The MOSFETs are also relatively sensitive to ESD, and can be damaged unintentionally through bad handling. The [TI DRV83xx](https://www.mouser.ca/c/semiconductors/power-management-ics/gate-drivers/?m=Texas%20Instruments&series=DRV8323) series chips that these controllers rely on have been heavily effected by the silicon shortage. You can see the Mouser lead time is May-July 2023, and I've seen as high as 2024 on Digikey. ODrive is also phasing these boards out in favour of their Pro line, so they are only while supplies last. The Odrive's usually take under a week to get there if you get the express shipping which was about $35USD to Toronto (worth it in my opinion). Otherwise it is about 2-3 weeks, which during the semester is a long time. 

If you are just building a test bench for the reaction wheels right now (not worried about integration yet), and there are no 24V supplies in the lab, I would recommend getting a [24V 15A PSU](https://www.amazon.ca/NEWSTYLE-Universal-Regulated-Switching-Computer/dp/B01CS5G8J6/ref=sr_1_6?crid=TSKOR2UXZ3S7&keywords=24V+power+supply&qid=1664057434&sprefix=24v+power+supply%2Caps%2C112&sr=8-6) from Amazon. Its cheap ($35.99) and could be there tomorrow. It's also good to have some [fork crimp connectors](https://www.amazon.ca/Glarks-240pcs-Terminals-Connectors-Assortment/dp/B01E5V5GMG/ref=sr_1_18?crid=36T7VDCQIH59A&keywords=u+crimp+connectors&qid=1664057987&sprefix=u+crimp+connectors%2Caps%2C100&sr=8-18) for that style of PSU and some wire of a reasonable gauge (12AWG is usually rated for about 20A), if they aren't already lying around the shop. Also, [these](https://www.amazon.ca/EKYLIN-Converter-Regulator-Adapter-Vehicle/dp/B01EFUHGMU) are a cheap easy way to step up the 12-24V and can source up to 10A.

**Comment by SeanTedesco** [comment]
There are a couple paths that I see we can go down: 

1) Continue with the RWs we currently have and determine a method to get 24v (30W) available on the satellite
2) Buy new RWs that are compatible with the current EPS module we have (which provides 3.3, 5.0 and 12-volt buses). 

Regarding option 1)
- PRO: less expensive, we did not budget for additional RWs this year (but we can readjust our budget)
- PRO: some work has already been put into working with these RWs, so this would not get lost.
- CON: designing such a high power boost converter, may be quite complex and generate quite a bit of heat. 

Regarding option 2)
- PRO: simpler design, less unnecessary power consumption.
- PRO: Maxon offers a nominal 9v (6W) version of the motors we have (and are the motors used in a [Master's Thesis for an ADCS design](https://core.ac.uk/download/pdf/37319831.pdf))
- CON: motors aint cheap, looking at ~$500 for 4x of [these maxon motors](https://www.maxongroup.com/maxon/view/content/cart). 

What we can do as next steps:
- [ ] Continue with current motors for testing control algorithms, hardware design, etc. 
- [ ] Determine what a high power boost converter circuit would look like (12v - 24) 
- [ ] Determine if we need to buy new reaction wheels

@Aidan-Sheedy what are your thoughts on this?

**Comment by Aidan-Sheedy** [comment]
Hey @SeanTedesco,

I think I agree with your plan. I don't think we should move to buying any new reaction wheel hardware until we are sure that we need to - at this point in the design process it would probably cause more headaches than it would solve.

Regarding the 24V nominal voltage - @anthonysteel Sean and I were going over the data sheet for the parts that we have and we can't find any reference to a 24V nominal voltage, the part numbers we have line up with this:

![Screenshot 2022-09-27 195127](https://user-images.githubusercontent.com/56231471/192657239-8a7cf9df-bd92-43c8-8229-74df40092a62.png)

**Comment by anthonysteel** [comment]
I gotcha, the part numbers in the [Maxon Reaction Wheel Guide](https://queensuca.sharepoint.com/teams/GROUP-QSET/_layouts/15/Doc.aspx?sourcedoc=%7B9fb96dfc-b12f-4372-a0a4-f91bb2467e9c%7D&action=edit&wd=target%28Satellite%20ADCS.one%7C1ca96692-dfb3-460b-88e6-898f0a9a2535%2FMaxon%20Reaction%20Wheel%20Guide%7C5212604b-afdd-421f-ac9c-da0e617bb9c5%2F%29&wdorigin=NavigationUrl) were 397172 (motor) and 466023 (controller). This was the [controller datasheet](https://www.maxongroup.com/medias/sys_master/root/8834400092190/466023-ESCON-Module-24-2-Hardware-Reference-En.pdf) I was referencing.

**Comment by Aidan-Sheedy** [comment]
Here's the Data sheet for the Maxon boards @alexnbartley:

[Controller DataSheet.pdf](https://github.com/queens-satellite-team/adcs/files/9691026/Controller.DataSheet.pdf)
[EC 45 flat DataSheet.pdf](https://github.com/queens-satellite-team/adcs/files/9691027/EC.45.flat.DataSheet.pdf)
